File: app/editor/registry.py
from __future__ import annotations

from collections.abc import Iterable

from app.editor.adapters import DEFAULT_ADAPTERS
from app.editor.adapters.base import BlockAdapter, InputKind


class BlockRegistry:
    def __init__(self, adapters: Iterable[BlockAdapter] = ()) -> None:
        self._by_type: dict[str, BlockAdapter] = {}
        self._canonical: dict[str, str] = {}
        for adapter in adapters:
            self.register(adapter)

    def register(self, adapter: BlockAdapter) -> None:
        if adapter.block_type in self._by_type:
            raise ValueError(f"Block adapter already registered: {adapter.block_type}")
        self._by_type[adapter.block_type] = adapter
        self._canonical[adapter.block_type] = adapter.block_type
        for alias in adapter.aliases:
            if alias in self._canonical:
                raise ValueError(f"Block alias already registered: {alias}")
            self._canonical[alias] = adapter.block_type

    def canonical_type(self, block_type: str) -> str:
        return self._canonical.get(block_type, block_type)

    def get(self, block_type: str) -> BlockAdapter | None:
        return self._by_type.get(self.canonical_type(block_type))

    def require(self, block_type: str) -> BlockAdapter:
        adapter = self.get(block_type)
        if adapter is None:
            raise KeyError(f"Unsupported block type: {block_type}")
        return adapter

    def supported_types(self) -> tuple[str, ...]:
        return tuple(self._by_type)

    def by_input_kind(self, input_kind: InputKind) -> tuple[str, ...]:
        return tuple(
            block_type
            for block_type, adapter in self._by_type.items()
            if adapter.input_kind == input_kind
        )

    def compatible_children(self, container_type: str) -> tuple[str, ...]:
        adapter = self.get(container_type)
        return adapter.child_types if adapter else ()


block_registry = BlockRegistry(DEFAULT_ADAPTERS)

__all__ = ["BlockRegistry", "InputKind", "block_registry"]

File: lib/editor/registry.js
export const InputKind = Object.freeze({
  TEXT: 'text',
  TELEGRAM: 'telegram',
  NATIVE: 'native',
  AUTOMATIC: 'automatic',
  CONTAINER: 'container',
});

export function blockAdapter({
  blockType,
  inputKind,
  aliases = [],
  childTypes = [],
  supportsCaption = false,
}) {
  return Object.freeze({
    blockType,
    inputKind,
    aliases: Object.freeze([...aliases]),
    childTypes: Object.freeze([...childTypes]),
    supportsCaption: Boolean(supportsCaption),
    get isContainer() { return inputKind === InputKind.CONTAINER; },
    validate(block) {
      if (![blockType, ...aliases].includes(String(block?.type ?? ''))) return [`expected ${blockType}`];
      if (!block?.data || typeof block.data !== 'object' || Array.isArray(block.data)) return ['block data must be a mapping'];
      return [];
    },
  });
}

const detailsChildren = [
  'paragraph','heading','preformatted','footer','divider','mathematical_expression','anchor',
  'list','blockquote','pullquote','table','collage','slideshow','map','animation','audio',
  'document','photo','video','voice',
];

export const DEFAULT_ADAPTERS = Object.freeze([
  blockAdapter({ blockType:'paragraph', inputKind:InputKind.TEXT, aliases:['text'] }),
  blockAdapter({ blockType:'heading', inputKind:InputKind.TEXT }),
  blockAdapter({ blockType:'preformatted', inputKind:InputKind.TEXT }),
  blockAdapter({ blockType:'footer', inputKind:InputKind.TEXT }),
  blockAdapter({ blockType:'divider', inputKind:InputKind.AUTOMATIC }),
  blockAdapter({ blockType:'mathematical_expression', inputKind:InputKind.NATIVE }),
  blockAdapter({ blockType:'anchor', inputKind:InputKind.TEXT }),
  blockAdapter({ blockType:'list', inputKind:InputKind.TEXT }),
  blockAdapter({ blockType:'blockquote', inputKind:InputKind.TEXT }),
  blockAdapter({ blockType:'pullquote', inputKind:InputKind.TEXT }),
  blockAdapter({ blockType:'collage', inputKind:InputKind.CONTAINER, childTypes:['photo','video'], supportsCaption:true }),
  blockAdapter({ blockType:'slideshow', inputKind:InputKind.CONTAINER, childTypes:['photo','video'], supportsCaption:true }),
  blockAdapter({ blockType:'table', inputKind:InputKind.TEXT }),
  blockAdapter({ blockType:'details', inputKind:InputKind.CONTAINER, childTypes:detailsChildren }),
  blockAdapter({ blockType:'map', inputKind:InputKind.TELEGRAM, supportsCaption:true }),
  blockAdapter({ blockType:'animation', inputKind:InputKind.TELEGRAM, supportsCaption:true }),
  blockAdapter({ blockType:'audio', inputKind:InputKind.TELEGRAM, supportsCaption:true }),
  blockAdapter({ blockType:'document', inputKind:InputKind.TELEGRAM, supportsCaption:true }),
  blockAdapter({ blockType:'photo', inputKind:InputKind.TELEGRAM, supportsCaption:true }),
  blockAdapter({ blockType:'video', inputKind:InputKind.TELEGRAM, supportsCaption:true }),
  blockAdapter({ blockType:'voice', inputKind:InputKind.TELEGRAM, supportsCaption:true }),
]);

export class BlockRegistry {
  constructor(adapters = []) {
    this.byType = new Map();
    this.canonical = new Map();
    for (const adapter of adapters) this.register(adapter);
  }

  register(adapter) {
    if (this.byType.has(adapter.blockType)) throw new Error(`Block adapter already registered: ${adapter.blockType}`);
    this.byType.set(adapter.blockType, adapter);
    this.canonical.set(adapter.blockType, adapter.blockType);
    for (const alias of adapter.aliases) {
      if (this.canonical.has(alias)) throw new Error(`Block alias already registered: ${alias}`);
      this.canonical.set(alias, adapter.blockType);
    }
  }

  canonicalType(blockType) {
    return this.canonical.get(blockType) ?? blockType;
  }

  get(blockType) {
    return this.byType.get(this.canonicalType(blockType)) ?? null;
  }

  require(blockType) {
    const adapter = this.get(blockType);
    if (!adapter) throw new Error(`Unsupported block type: ${blockType}`);
    return adapter;
  }

  supportedTypes() {
    return [...this.byType.keys()];
  }

  byInputKind(inputKind) {
    return [...this.byType.entries()]
      .filter(([, adapter]) => adapter.inputKind === inputKind)
      .map(([blockType]) => blockType);
  }

  compatibleChildren(containerType) {
    return [...(this.get(containerType)?.childTypes ?? [])];
  }
}

export const blockRegistry = new BlockRegistry(DEFAULT_ADAPTERS);
